src/main/java/org/opensearch/flowframework/indices/FlowFrameworkIndicesHandler.java

```diff
@@ -716,26 +679,93 @@ public void addResourceToStateIndex(
+        String workflowId,
+        ResourceCreated newResource,
+        int retries,
+        ActionListener<WorkflowData> listener
+    ) {
+        GetRequest getRequest = new GetRequest(WORKFLOW_STATE_INDEX, workflowId);
+        client.get(getRequest, ActionListener.wrap(getResponse -> {
+            if (!getResponse.isExists()) {
+                listener.onFailure(new FlowFrameworkException("Workflow state not found for " + workflowId, RestStatus.NOT_FOUND));
+                return;
+            }
+            WorkflowState currentState = WorkflowState.parse(getResponse.getSourceAsString());
+            List<ResourceCreated> resourcesCreated = new ArrayList<>(currentState.resourcesCreated());
+            resourcesCreated.add(newResource);
+            XContentBuilder builder = XContentFactory.jsonBuilder();
+            WorkflowState newState = WorkflowState.builder(currentState).resourcesCreated(resourcesCreated).build();
+            newState.toXContent(builder, null);
+            UpdateRequest updateRequest = new UpdateRequest(WORKFLOW_STATE_INDEX, workflowId).doc(builder)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setIfSeqNo(getResponse.getSeqNo())
+                .setIfPrimaryTerm(getResponse.getPrimaryTerm());
+            client.update(
+                updateRequest,
+                ActionListener.wrap(
+                    r -> handleStateUpdateSuccess(workflowId, newResource, listener),
+                    e -> handleStateUpdateException(workflowId, newResource, retries, listener, e)
+                )
             );
-        } catch (Exception e) {
-            String errorMessage = "Failed to parse and update new created resource";
-            logger.error(errorMessage, e);
-            listener.onFailure(new FlowFrameworkException(errorMessage, ExceptionsHelper.status(e)));
+        }, ex -> handleStateUpdateException(workflowId, newResource, 0, listener, ex)));
+    }
+
+    private void handleStateUpdateSuccess(String workflowId, ResourceCreated newResource, ActionListener<WorkflowData> listener) {
+        String resourceName = newResource.resourceType();
+        String resourceId = newResource.resourceId();
+        String nodeId = newResource.workflowStepId();
+        logger.info("Updated resources created for {} on step {} with {} {}", workflowId, nodeId, resourceName, resourceId);
+        listener.onResponse(new WorkflowData(Map.of(resourceName, resourceId), workflowId, nodeId));
+    }
+
+    private void handleStateUpdateException(
+        String workflowId,
+        ResourceCreated newResource,
+        int retries,
+        ActionListener<WorkflowData> listener,
+        Exception e
+    ) {
+        if (e instanceof VersionConflictEngineException && retries > 0) {
+            // Retry if we haven't exhausted retries
+            getAndUpdateResourceInStateDocumentWithRetries(workflowId, newResource, retries - 1, listener);
```

Comment:
Why this recursive calling here when we can pass retryonconflict?

Reply:
RetryOnConflict only applies when you don't care what the existing value is but you're updating a field.  In our case we need to fetch the existing resource list, add to it, and then put that resource list back.  

If the document changed in between us getting and updating, retryOnConflict would miss it.  We use the `seqNo` and `primaryTerm` from the document we "got" to make sure that's the same document we're putting back in; if the doc has been updated, retrying won't do us any good, we need to retry the whole get-and-put which is what this does.

Also a minor note, while it looks recursive, it's just a chained callback not unlike all our other transport calls, so it won't fill up the stack.